### PR TITLE
CV2 implementation

### PR DESCRIPTION
**Pull request by msaggiorato**
This PR contains a lot of work done throughout the last 4/5 months implementing Checkout V2.

I assume @vbelolapotkov will be the code reviewer here.

If you want to review this in a more staged way (smaller chunks) we do have smaller individual PRs in our fork repo.

For the time being I'm creating this as draft, but as soon as we agree on the review process we can turn this into a non draft PR.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -40,7 +39,7 @@
     "watchsass": "node_modules/.bin/node-sass $npm_package_assets_styles_sass --output $npm_package_assets_styles_css --output-style compressed --watch",
     "postsass": "for f in $npm_package_assets_styles_css; do file=${f%.css}; node_modules/.bin/cleancss -o $file.css $f; done",
     "makepot": "wpi18n makepot --domain-path languages --pot-file woocommerce-gateway-amazon-payments-advanced.pot --type plugin --main-file woocommerce-gateway-amazon-payments-advanced.php --exclude node_modules,tests,docs",
```
Comment (vbelolapotkov) [inline package.json:41]: Exclude docker when generating translations file.
```suggestion
    "makepot": "wpi18n makepot --domain-path languages --pot-file woocommerce-gateway-amazon-payments-advanced.pot --type plugin --main-file woocommerce-gateway-amazon-payments-advanced.php --exclude node_modules,tests,docs,docker",
```

File: package.json [hunk 2]
```diff
@@ -40,7 +39,7 @@
...
-    "woorelease": "npm run build"
```
Comment (vbelolapotkov) [inline package.json:43]: `woorelease` is not script is not used anymore, `build` script used instead to prepare plugin package for release, so some adjustments are needed to scripts:

1. Move current build script into `build:assets`.
2. Make build script to look like this: `"build": "npm run build:assets && npm run archive",`

**Review by vbelolapotkov — CHANGES_REQUESTED** [review]
Added a couple of comments about scripts changes which will be needed to package and release the plugin.

File: package.json [hunk 3]
```diff
@@ -30,7 +30,6 @@
     "test:single": "cross-env NODE_CONFIG_DIR='./tests/e2e/config' BABEL_ENV=commonjs mocha --compilers js:babel-register",
     "update-deps": "npm install -g rimraf && rimraf node_modules && rimraf npm-shrinkwrap.json && npm install && npm shrinkwrap --dev",
     "build": "npm run uglify && npm run makepot && npm run sass",
```
Comment (kalessil) [inline package.json:32]: First feedback on build artifact produced:
 - .idea, phpstan.neon needs to be excluded (some folks are using JB IDEs and .idea sneaks into the build artifact)
 - vendor folder was missing, the build task should build PHP/JS production dependencies

Let's push the related changes in this PR.

Comment (msaggiorato) [reply package.json:32]: Hey @kalessil , about these:

* Ok to ignore these, i'd say these goes to the composer.json file.
* About vendor folder not being present, i didn't change how the build process was from before our work here. I suppose i should do a `composer install --production` before the `composer archive` command?

Thanks

Comment (kalessil) [reply package.json:32]: Hey @msaggiorato,
- Sound good
- Yep. I'd suggest also wrapping it into a dedicated command like `"build:deps": "rm -rf vendor && composer install --no-dev --optimize-autoloader",`

**Comment by kalessil** [comment]
For a better visibility, sharing testing scenarios (WIP, please ignore while I'm continue testing):
- Regular purchase (authorize and capture) in USD  :heavy_check_mark: 
- Regular purchase (authorize and capture) in EUR
- Regular purchase (authorize, but capture manually) in USD :heavy_check_mark: 
- Refund order in USD  :heavy_check_mark: 
- Refund order in EUR
- Subscription purchase (regular)  :heavy_check_mark: 
- Subscription purchase (multiple) 
- Subscription purchase (free trial)
- Subscription renew
- Subscription payment method change
- Subscription cancellation

Findings:
- the payment button with 
Screenshot from 2021-04-28 11-48-26

**Comment by kalessil** [comment]
WIP: architecture and compatibility review.

- Review architecture changes
- Test upgrade/downgrade scenario for the plugin

**Comment by msaggiorato** [comment]
> Architecture and compatibility review.
> 
> * Review architecture changes ✔️
>   
>   * overall looks good (similar to the original, I also like how  legacy and compatibility handling is organized) ✔️
>   * also thanks for refactoring some of the complex code fragments in v2 (e.g. order actions handling, logging, and more) 🥇
> * Test upgrade/downgrade scenario for the plugin
>   
>   * upgrade 🛑
>     
>     * the release artifact name is incorrect and considered to be a new plugin causing class declaration conflicts
>     * `"name": "woocommerce-gateway-amazon-payments-advanced"` in package.json solves this specific problem
>   * downgrade 🛑
>     
>     * after the downgrade, I don't see any hints about re-connecting Amazon account, as I saw on the upgrade
>     * purchases are getting broken (see screenshot), as I'm not aware of needed re-connection
> 
> Screenshot from 2021-04-29 10-49-41

* The slug of the plugin changed on the original version as well. Just left it the same. Any problem with upgrades related to that is already present on 1.13 (at least).
* Regarding downgrade. That error should not be there. The way we tested is by having the 2 versions of the plugin on the site (with folder names changed), and enabling one or the other. of course orders made with the v2 api will not be able to be processed by the old plugin, since data changed. But v1 orders can still be processed. And definetely new orders. How did you test that?

**Comment by msaggiorato** [comment]
> For a better visibility, sharing testing scenarios:
> 
> * Regular purchase (authorize and capture) in USD  ✔️
> * Regular purchase (authorize and capture) in EUR  ✔️
> * Regular purchase (authorize, but capture manually) in USD ✔️
> * Refund order in USD  ✔️
> * Refund order in EUR  ✔️
> * Subscription purchase (regular)  ✔️
> * Subscription purchase (multiple) ✔️
> * Subscription purchase (free trial) ✔️
> * Subscription renew ✔️
> * Subscription payment method change 🛑
> * Subscription cancellation ✔️
> 
> Findings:
> 
> * the payment button width is not aligned with checkout button
>   Screenshot from 2021-04-28 11-48-26
> * nothing happens when clicking 'Change' while changing an existing subscription payment method
>   Screenshot from 2021-04-28 13-36-17
>   Note: tested with the latest stable version of Firefox.

* The issue with the cart, is a limitation of the Amazon widget. We do not set any width limit, but it'll limit itself to 450 px wide. We've never seen that case with storefront tho, which page is that? Or which viewport?
* Will definetely check the change payment method flow. That was working.

**Comment by kalessil** [comment]
> The issue with the cart, is a limitation of the Amazon widget. We do not set any width limit, but it'll limit itself to 450 px wide. We've never seen that case with storefront tho, which page is that? Or which viewport?

It's exactly 450px, env is a regular browser.

> Will definitely check the change payment method flow. That was working.

Let me know if you can reproduce it and if I can provide more details here.

**Comment by kalessil** [comment]
> The slug of the plugin changed on the original version as well. Just left it the same. Any problem with upgrades related to that is already present on 1.13 (at least).

> Regarding downgrade. That error should not be there. The way we tested is by having the 2 versions of the plugin on the site (with folder names changed), and enabling one or the other. of course orders made with the v2 api will not be able to be processed by the old plugin, since data changed. But v1 orders can still be processed. And definetely new orders. How did you test that?

My steps were:
- Download and install `woocommerce-gateway-amazon-payments-advanced.1.13.1.zip`
- Connect Amazon account, do test purchases
- Build and install `woocommerce-gateway-amazon-payments-advanced.zip` (my expectation was that plugin slug will be the same)
- Do test piurchase
- Re-connect Amazon account, do test purchase.
- Download and install `woocommerce-gateway-amazon-payments-advanced.1.13.1.zip`
- Do test purchases

If I understand correctly, the key point here is that we'll have 2 different slags and upgrade/downgrade is basically switching from one slug to another. Is that right?

**Comment by msaggiorato** [comment]
Hey @kalessil [here's](https://github.com/woocommerce/woocommerce-gateway-amazon-pay/commit/1121d389c80546a3d123843a4c32a5c74e8f73f7) the commit that changed the slug, i just updated the package json in our file (which we based from a previous state) to reflect this.

We find it strange also that the slug of the plugin changed at some point, but you made that decision on your side, and I didn't want to contest that.

**Comment by msaggiorato** [comment]
> Let me know if you can reproduce it and if I can provide more details here.

Cannot reproduce here, can you provide more details?

File: package.json (hunk 3 above)
Comment (msaggiorato) [reply package.json]: Resolved

File: package.json (hunk 2 above)
Comment (msaggiorato) [reply package.json:43]: Resolved

File: package.json (hunk 1 above)
Comment (msaggiorato) [reply package.json]: Added

**Comment by kalessil** [comment]
> Hey @kalessil here's the commit that changed the slug, i just updated the package json in our file (which we based from a previous state) to reflect this.

Oh, I see now. No worries I'll try the upgrade path based on the slug change then.

> Cannot reproduce here, can you provide more details?

Ok, will do. I want to do it on a clean system though, to ensure it's not something caused by me.

**Comment by kalessil** [comment]
@msaggiorato : thanks for the fixes :+1:.

- The issue with the non-working 'Change' link in the subscription method change has gone.
- Amazon Pay button centering look good.

The only topic left is downgrade: it's still not working.

**Review by kalessil — APPROVED** [review]

**Comment by kalessil** [comment]
@msaggiorato : pre-approving the PR. Once the tests are fixed, I'll merge the PR.

**Pull request merged**